DEVELOPMENT_BEST_PRACTICE.md

```diff
@@ -0,0 +1,194 @@
+
+##### Variable
+
+Variable names should be concise and descriptive. Prefer single word names. The further away from the site of declaration a variable is used, the more descriptive it needs to be. For example, a variable named `n` used 30 lines after declaration makes it unnecessarily difficult to reason what it represents at the site of use.
+
+### Abstractions
+
+- Functions & Methods
+- Constructs
+- Interfaces
+
+### <a name="dependencies"></a> Dependencies
+
+- Accept interfaces, return structs. Where should we put this note? We ened to clarify it as its misrepresentative without explanation hence why it doesn't feature in Go proverbs and has been explicitly rejected by the Go team.
+
+Dependencies exist in various forms. Often, developers talk about object dependencies derived from object oriented programming. However, dependencies exist in functions and even packages. Namely, a function accepts parameters, therefore all parameters it accepts are dependencies of the function. Similarly, every import statement (other than those in the standard library given the standard library is part of the language) in a soruce file represents a dependency a package has.
+
+Dependencies should be delcared explicitly. Go source files do this naturally with explicit import statements. Functions should explicitly declare their dependencies as parameters thereby honoring the [dependency inversion principle](https://www.digitalocean.com/community/conceptual_articles/s-o-l-i-d-the-first-five-principles-of-object-oriented-design#dependency-inversion-principle).
+
+```go
+func FastenPlanks(tools Toolbox, surface Surface, planks Planks) error {
+    var plank Plank
+    for ; planks.Count() > 0; plank = planks.Next() {
+        if err := tools.Hammer(surface, ); err != nil {
+            return err
+        }
+    }
+}
+```
+
+Avoid constructing dependencies in-line. If you do construct a dependency in-line because it compliments default usage, provide a way to override it such as with a setter.
+
+#### Loggers
+
+Loggers are sometimes treated as a special depency by making entire projects depend on a package level logger. This _masks_ a crucial dependency that is hard to change without impacting the rest of the code base. To understand why this is bad, consider extracting a package into its own repository  for reusability. Depending on a package level logger, or any other package level dependencies, makes the extraction process unnecessarily difficult as you first need to pull the hidden dependency apart.
+
+Logger's are not special dependencies, inject them like any other dependency. If you want to minimize the surface area of the default initializing constructor (`New...` funcs) you may construct noop loggers to satisfy the logger dependency and provide a means to override the logger with a setter.
```

Comment:
"inject" makes me think about dependency injection. Youre not suggested that here are you? might think about a synonym?

Reply:
That's exactly what I'm suggesting :).

Reply:
See https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_package_level_state and

Reply:
Ah yes. I'm very much on the same page here. I believe my previous definition of dependency injection was incorrect.